File: back/src/api/fetch_titles.rs
use std::sync::Arc;

use crate::{
    domain::{models::Title, use_cases},
    infra::TitleRepository,
};
use axum::{http::StatusCode, Extension, Json};
use serde::Serialize;

#[derive(Serialize)]
pub struct Response {
    title: String,
}

impl From<Title> for Response {
    fn from(title: Title) -> Self {
        Self {
            title: title.into(),
        }
    }
}

type Error = (StatusCode, &'static str);

pub async fn fetch_titles(
    Extension(repo): Extension<Arc<dyn TitleRepository>>,
) -> Result<Json<Vec<Response>>, Error> {
    match use_cases::fetch_titles(repo).await {
        Ok(titles) => Ok(titles
            .into_iter()
            .map(|title| title.into())
            .collect::<Vec<Response>>()
            .into()),
        Err(use_cases::fetch_titles::Error::Unknown(str)) => {
            Err((StatusCode::INTERNAL_SERVER_ERROR, str))
        }
    }
}

File: src/domain/lib/fetchTitles.ts
import { Result } from "@utils";
import { TitleRepository } from "@infra";
import { Title } from "../models/mod.ts";

export async function fetchTitles(titleRepository: TitleRepository): Promise<Result<Title[], string>> {
    return Result.match(await titleRepository.fetchAll(), {
        onOk: (titles) => Result.ok(titles),
        onErr: (err) => Result.err(err)
    });
}